test/bio_socket_test.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <fcntl.h>
+#include <openssl/bio.h>
+#include <netinet/tcp.h>
+
+#include "testutil.h"
+
+#if !defined(OPENSSL_NO_SOCK)
+
+/* 
+ * based on helpers/ssltestlib.c:create_test_sockets()
+ * IPv4-only
+ */
+
+#define SOCKET_DATA "FooBar"
+#define SOCKET_DATA_LEN sizeof(SOCKET_DATA)
+
+/* 
+ * This has to be run twice, the first attempt will create
+ * the fast open cookie, this can be seen in `ip tcp_metrics`
+ * and in /proc/net/netstat/ on Linux. e.g. on Linux 4.15.0-135-generic:
+ * $ grep '^TcpExt:' /proc/net/netstat | cut -d ' ' -f 84-90 | column -t
+ * The second attempt will use the cookie and actually do TCP fast open.
+ */
+static int test_bio_socket(int idx)
+{
+    struct sockaddr_in sin;
+    socklen_t slen = sizeof(sin);
+    const char *host = "127.0.0.1";
+    int cfd_connected = 0;
+    int ret = 0;
+    int cfd = -1; /* client socket */
+    int afd = -1; /* accept socket */
+    int sfd = -1; /* server accepted socket */
+    BIO_ADDR *baddr = NULL;
+    int data_written = 0;
+    char read_buffer[20];
+    int bytes_read;
+
+    /* ADDRESS SETUP */
+    memset(&sin, 0, sizeof(sin));
+    sin.sin_family = AF_INET;
+    sin.sin_addr.s_addr = inet_addr(host);
+
+    baddr = BIO_ADDR_new();
+    if (!TEST_ptr(baddr))
+        goto err;
+    if (!TEST_true(BIO_ADDR_rawmake(baddr, sin.sin_family, &sin.sin_addr.s_addr,
+                                    sizeof(sin.sin_addr.s_addr), sin.sin_port)))
+        goto err;
+
+    /* ACCEPT SOCKET */
+    afd = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
Why not BIO_socket?

Reply:
WIP; I based this on other OpenSSL code that used vanilla socket... I'm slowly moving to use `BIO_` calls as I refine the test.

Reply:
Fixed